File: gable/src/main/java/com/nhaarman/gable/container/LinearLayoutContainer.java
package com.nhaarman.gable.container;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import butterknife.ButterKnife;
import com.nhaarman.gable.presenter.Presenter;
import org.jetbrains.annotations.NotNull;

public abstract class LinearLayoutContainer<P extends Presenter<P, C>, C extends Container<P, C>>
    extends LinearLayout
    implements Container<P, C> {

  private P mPresenter;

  protected LinearLayoutContainer(final Context context) {
    super(context);
  }

  protected LinearLayoutContainer(final Context context, final AttributeSet attrs) {
    super(context, attrs);
  }

  protected LinearLayoutContainer(final Context context, final AttributeSet attrs, final int defStyle) {
    super(context, attrs, defStyle);
  }

  @NotNull
  public P getPresenter() {
    if (mPresenter == null) {
      throw new NullPointerException("Presenter has not been set.");
    }

    return mPresenter;
  }

  @Override
  public void setPresenter(@NotNull final P presenter) {
    mPresenter = presenter;
  }

  @Override
  protected void onFinishInflate() {
    super.onFinishInflate();
    if (isInEditMode()) {
      return;
    }

    ButterKnife.inject(this);
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    if (isInEditMode()) {
      return;
    }

    getPresenter().acquire((C) this);
  }

  @Override
  protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();
    getPresenter().releaseContainer();
  }
}

File: triad-kotlin/src/main/kotlin/com/nhaarman/triad/LinearLayoutContainer.kt
/*
 * Copyright 2016 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nhaarman.triad

import android.content.Context
import android.util.AttributeSet
import android.widget.LinearLayout

abstract class LinearLayoutContainer<P : Presenter<*, ActivityComponent>, ActivityComponent> @JvmOverloads constructor(context: Context, attrs: AttributeSet?, defStyle: Int = 0) : LinearLayout(context, attrs, defStyle), Container {

    val presenter: P by lazy { findPresenter<P>(context, this) }

    val activityComponent: ActivityComponent by lazy { findActivityComponent<ActivityComponent>(context) }

    @Suppress("UNCHECKED_CAST")
    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        if (isInEditMode) {
            return
        }

        (presenter as Presenter<Container, ActivityComponent>).acquire(this, activityComponent)
    }

    @Suppress("UNCHECKED_CAST")
    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        (presenter as Presenter<Container, ActivityComponent>).releaseContainer(this)
    }

    override final fun context() = context
}
